pkg/instancetype/instancetype.go

```diff
@@ -510,31 +509,134 @@ func (m *methods) getInstancetypeSpecRevision(revisionName string, namespace str
 	return &instancetypeSpec, nil
 }
 
-func (m *methods) findInstancetype(vm *virtv1.VirtualMachine) (*instancetypev1alpha1.VirtualMachineInstancetype, error) {
-
-	if vm.Spec.Instancetype == nil {
-		return nil, nil
+func (m *methods) findInstancetype(name, namespace string) (*instancetypev1alpha1.VirtualMachineInstancetype, error) {
+	instancetype, err := m.clientset.VirtualMachineInstancetype(namespace).Get(context.Background(), name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
 	}
+	return instancetype, nil
+}
 
-	instancetype, err := m.clientset.VirtualMachineInstancetype(vm.Namespace).Get(context.Background(), vm.Spec.Instancetype.Name, metav1.GetOptions{})
+func (m *methods) findClusterInstancetype(name string) (*instancetypev1alpha1.VirtualMachineClusterInstancetype, error) {
+	instancetype, err := m.clientset.VirtualMachineClusterInstancetype().Get(context.Background(), name, metav1.GetOptions{})
 	if err != nil {
 		return nil, err
 	}
 	return instancetype, nil
 }
 
-func (m *methods) findClusterInstancetype(vm *virtv1.VirtualMachine) (*instancetypev1alpha1.VirtualMachineClusterInstancetype, error) {
-
-	if vm.Spec.Instancetype == nil {
+func (m *methods) FindPreferredInstancetype(vm *virtv1.VirtualMachine) (*virtv1.InstancetypeMatcher, error) {
+	if vm.Spec.Instancetype != nil || len(vm.Spec.Template.Spec.Volumes) == 0 {
 		return nil, nil
 	}
 
-	instancetype, err := m.clientset.VirtualMachineClusterInstancetype().Get(context.Background(), vm.Spec.Instancetype.Name, metav1.GetOptions{})
-	if err != nil {
-		return nil, err
+	var (
+		preferredInstancetype     string
+		preferredInstancetypeKind string
+		hasAnnotation             bool
+		hasKindAnnotation         bool
+		instancetype              *instancetypev1alpha1.VirtualMachineInstancetype
+		clusterInstancetype       *instancetypev1alpha1.VirtualMachineClusterInstancetype
+		instancetypeSpec          instancetypev1alpha1.VirtualMachineInstancetypeSpec
+	)
+
+	// Kind defaults to cluster as it does when users provide a InstancetypeMatcher within the VirtualMachine
+	preferredInstancetypeKind = apiinstancetype.ClusterSingularResourceName
+
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
```

Comment:
what if multiple pvcs are present and have preferred instance types or preferred preference type annotations.

I know that would be rare, but it could happen if we're using a VM to modify an image for another VM, or something weird like that. basically it's not out of the realm of possibilities.

Reply:
Yeah definitely, I'd likely document the current behaviour where the first set of annotations are used, assuming the list of volumes always remains consistent and thus making the outcome reproducible.